app/src/org/commcare/preferences/DeveloperPreferences.java

```diff
@@ -215,9 +235,33 @@ public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, Strin
                             .apply();
                 }
                 break;
+            case USER_ENTERED_ACCESS_PIN:
+                if (userEnteredAccessPinMatches()) {
+                    getPreferenceScreen().removeAll();
+                    initAllPrefs();
+                    Toast.makeText(this, Localization.get("dev.options.access.granted"),
+                            Toast.LENGTH_SHORT).show();
+                } else {
+                    clearUserEnteredAccessPin();
+                    Toast.makeText(this, Localization.get("dev.options.pin.incorrect"),
+                            Toast.LENGTH_SHORT).show();
+                }
+                break;
         }
     }
 
+    private static boolean userEnteredAccessPinMatches() {
+        SharedPreferences prefs = CommCareApplication.instance().getCurrentApp().getAppPreferences();
+        String projectAccessPin = prefs.getString(PROJECT_SET_ACCESS_PIN, null);
+        String userAccessPin = prefs.getString(USER_ENTERED_ACCESS_PIN, null);
+        return userAccessPin == null ? false : userAccessPin.equals(projectAccessPin);
+    }
+
+    private static void clearUserEnteredAccessPin() {
```

Comment:
We should fire this event when the user de-activates developer mode as well, I think. Otherwise the PIN will still be set and will allow the keystroke pattern to re-active developer mode.

Reply:
And by "de-activate developer mode", I'm assuming you mean editing the first setting available in Developer Preferences? Otherwise I don't know what other mechanism there is for this. But I never really imagined anyone actually using that setting to turn _off_ access.

Reply:
You'd want to disable to return the device to "production" settings.

It's definitely part of the workflow if you used something from developer settings in the field.

 * Admin is in the field with the phone and user
 * Turns on Developer mode to access / change something
 * makes the change
 * Turns off developer mode afterward so they can give the device back without granting access to those features.